rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp

```diff
@@ -531,13 +532,39 @@ void SqliteStorage::prepare_for_reading()
         topic_list += ",";
       }
     }
-    where_conditions.push_back("(topics.name IN (" + topic_list + "))");
+    topic_and_service_list = "(topics.name IN (" + topic_list + "))";
   }
-  // add topic filter based on regular expression
-  if (!storage_filter_.topics_regex.empty()) {
+
+  // add service filter
+  if (!storage_filter_.services.empty()) {
     // Construct string for selected topics
-    where_conditions.push_back("(topics.name REGEXP '" + storage_filter_.topics_regex + "')");
+    std::string service_list{""};
+    for (auto & service : storage_filter_.services) {
+      service_list += "'" + service + "'";
+      if (&service != &storage_filter_.topics.back()) {
+        service_list += ",";
+      }
+    }
+
+    topic_and_service_list = topic_and_service_list +
+      std::string(topic_and_service_list.empty() ? "" : " OR ") +
+      "(topics.name IN (" + service_list + "))";
```

Comment:
Comments from MichaelOrlov

Consider using where_conditions.push_back(..) for each condition instead of gluing a long string with OR conditions.

Reply:
@MichaelOrlov 

push_back() only adds one character.  I don't know how to change.   
I change like the below for readability. What do you think?
```c++
    if (!topic_and_service_list.empty()) {
      topic_and_service_list.append(" OR ");
    }
    topic_and_service_list.append("(topics.name IN (" + service_list + "))");
```

Reply:
@Barry-Xu-2018 I meant conceptually instead of trying to glue a long `topics_and_service_list` string with multiple `OR`  and another condition. e.g `append("OR (topics.name IN (" + service_list + "))")` in loop.  We can add each condition without **OR** using 
```
where_conditions.push_back("(topics.name IN (" + service_list + "))");
```
in the loop.  
Your assumption that `push_back() only adds one character.` is incorrect.
Look how logic was originally:
```
   where_conditions.push_back("(topics.name IN (" + topic_list + "))");
  }
  // add topic filter based on regular expression
  if (!storage_filter_.topics_regex.empty()) {
    // Construct string for selected topics
    where_conditions.push_back("(topics.name REGEXP '" + storage_filter_.topics_regex + "')");
  }
  // exclude topics based on regular expressions
  if (!storage_filter_.topics_regex_to_exclude.empty()) {
    // Construct string for selected topics
    where_conditions.push_back(
      "(topics.name NOT IN "
      "(SELECT topics.name FROM topics WHERE topics.name REGEXP '" +
      storage_filter_.topics_regex_to_exclude + "'))");
  }
```